### PR TITLE
Add openmetrics and exemplars support

### PR DESCRIPTION
**Pull request by voltbit**
example

### Overview

- Added support for the OpenMetrics standard
- Added support for automatically providing exemplars populated with trace information from OpenTelemetry

The main purpose of the PR is to enable the use of Exemplars for NodeJS codebase.

Relevant resources:

[Grafana exemplars](https://grafana.com/docs/grafana/latest/basics/exemplars/)
[Prometheus exemplars](https://prometheus.io/docs/prometheus/latest/feature_flags/#exemplars-storage)
[OpenMetics spec](https://github.com/OpenObservability/OpenMetrics/blob/d99b705f611b75fec8f450b05e344e02eea6921d/specification/OpenMetrics.md)
[Prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/)

### Design

The OpenTelemetry standard is very close to the original Prometheus format. In order to keep the library as backwards compatible as possible the default format is kept unchanged (Prometheus) and the use of exemplars is disabled.

The new features can be toggled:

- The format at the registry level (prometheus/openmetrics)
- The exemplars at the metric level

Each registry instance has an attribute (contentType) that will decide the format.
The two possible formats are defined by the constants `OPENMETRICS_CONTENT_TYPE` and `PROMETHEUS_CONTENT_TYPE` which contain the HTTP content type.
Future versions [should default](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#protocol-negotiation) to the `1.0.0` version.

Each metric has a flag for enabling the exemplar, the flag is put on the metrics supeclass for simplicity, but out of the currently implemented metric types only histograms and counters can have [exemplars](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#metric-types-1).

The biggest change to the code is the creation of separate functions for Counter increment and Histogram observe. Because the functions need to support a third optional parameter (exemplar labels) I have changed the way parameters are passed to the functions. Instead of using plain `(label, value)` the users will need to provide a single object with the format `({labels, value, exemplarLabels})`.
The change should not impact existing users, but users who want to use exemplars will need to use the new call format.

**Exemplar object**

Timestamp - is the time when the exemplar was created
Reference from the Golang client: [https://github.com/prometheus/client_golang/blob/1b145cad6847a692bd07e872d64b7102d33213c6/prometheus/histogram.go#L432](https://github.com/prometheus/client_golang/blob/1b145cad6847a692bd07e872d64b7102d33213c6/prometheus/histogram.go#L432).

[There is a hard 128 UTF-8 character limit on exemplar length.](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#size-limits)

The labels use for out of the box traces are `traceId` and `spanId`, it feels more like JavaScript to me, there is no other reason for the name choice. The golang implementation seems to be using `traceID` [here](https://github.com/prometheus/client_golang/blob/master/api/prometheus/v1/api_test.go#L1247) and the Java impl. uses `trace_id` [here](https://github.com/prometheus/client_java/blob/master/simpleclient/src/main/java/io/prometheus/client/exemplars/DefaultExemplarSampler.java#L12-L13). The label used for exemplars can be changed in Grafana. 

### Counters in OpenMetrics

Counters have a brekaing change in the form of an enforced `_total` suffix, it is not just a convention anymore. Examples:

**Prometheus**

```
# HELP mycounter help
# TYPE mycounter counter
mycounter 0

# HELP mycounter2_total help
# TYPE mycounter2_total counter
mycounter2_total 0
```

**OpenMetrics**

```
# HELP mycounter help
# TYPE mycounter counter
mycounter_total 0
```

Prometheus ignores the comments related to name and type, but the name of the metrics changes too and has the potential to break dashboards/alerts etc. The current implementation follows the same approach as the Java implementation here:

[https://github.com/prometheus/client_java/blob/master/simpleclient/src/main/java/io/prometheus/client/Counter.java#L72-L108](https://github.com/prometheus/client_java/blob/master/simpleclient/src/main/java/io/prometheus/client/Counter.java#L72-L108)

However, instead of applying the suffix at the level of the Counter object, this implementation applyes the change in the Registry object. The disadvantage is that the code is less elegant. The advantage is that the change is not breaking in any way for the existing users - the `_total` suffix will only be enforced by OpenMetrics registries, not Prometheus registries.

In the future, when OpenMetrics becomes more widely adopted, the behaviour can be moved inside Counter object and made mandatory.

---

### Benchmarks

Benchmark tests were not changed. They are using the default registry type (Prometheus) and no exemplars, so it is a check to see the impact for current users of the library. Ran 4 tests (results in gists bellow). The highest impact was on the registry benchmark with a ~10-15% performance hit.

```
⚠ registry ➭ getMetricsAsJSON#1 with 64 is 5.345% acceptably slower.
⚠ registry ➭ getMetricsAsJSON#2 with 8 is 3.076% acceptably slower.
⚠ registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 is 4.063% acceptably slower.
✓ registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 is 0.1468% faster.
⚠ registry ➭ getMetricsAsJSON#6 with 2 is 4.137% acceptably slower.
✗ registry ➭ metrics#1 with 64 is 11.50% slower.
✓ registry ➭ metrics#2 with 8 is 0.2174% faster.
⚠ registry ➭ metrics#2 with 4 and 2 with 2 is 1.046% acceptably slower.
⚠ registry ➭ metrics#2 with 2 and 2 with 4 is 3.413% acceptably slower.
✗ registry ➭ metrics#6 with 2 is 15.03% slower.
⚠ histogram ➭ observe#1 with 64 is 0.2735% acceptably slower.
⚠ histogram ➭ observe#2 with 8 is 0.5325% acceptably slower.
⚠ histogram ➭ observe#2 with 4 and 2 with 2 is 0.09087% acceptably slower.
⚠ histogram ➭ observe#2 with 2 and 2 with 4 is 0.3243% acceptably slower.
⚠ histogram ➭ observe#6 with 2 is 0.6302% acceptably slower.
✓ gauge ➭ inc is 16.96% faster.
⚠ gauge ➭ inc with labels is 1.991% acceptably slower.
⚠ summary ➭ observe#1 with 64 is 2.261% acceptably slower.
✓ summary ➭ observe#2 with 8 is 2.166% faster.
✓ summary ➭ observe#2 with 4 and 2 with 2 is 0.4552% faster.
⚠ summary ➭ observe#2 with 2 and 2 with 4 is 1.407% acceptably slower.
⚠ summary ➭ observe#6 with 2 is 1.116% acceptably slower.
```

[https://gist.github.com/voltbit/55bfdafccb5a0458d0b2aff9703dae43](https://gist.github.com/voltbit/55bfdafccb5a0458d0b2aff9703dae43)
[https://gist.github.com/voltbit/1e1097e6400638334e11da52fefcd5d4](https://gist.github.com/voltbit/1e1097e6400638334e11da52fefcd5d4)
[https://gist.github.com/voltbit/41828df848a7132c1aad196414ea2d69](https://gist.github.com/voltbit/41828df848a7132c1aad196414ea2d69)
[https://gist.github.com/voltbit/70539929453a2b95d4f9ac2df6707a9b](https://gist.github.com/voltbit/70539929453a2b95d4f9ac2df6707a9b)

_**TODO**_

- [x] Complete test coverage of the new features
- [x] Performance tests
- [x] Add better examples and readme info
- [x] ~~Strategy for registry merge when there are different registry formats~~
   -  decided to consider the merge of two different types of registries undefined behaviour, the users should always use the same type - Prometheus or OpenMetrics - if merging
- [x] Hadling the `_total` suffix on counters

**Not implemented**

- Support for the `_created` suffix on any metrics

### CODE REVIEW

**Comment by SimenB** [comment]
Very exciting, thanks for working on this!

File: example/server.js [hunk 1]
```diff
@@ -91,6 +92,8 @@ setInterval(() => {
 
 server.get('/metrics', async (req, res) => {
 	try {
+		// register.contentType = Registry.OPENMETRICS_CONTENT_TYPE;
+		register.contentType = Registry.PROMETHEUS_CONTENT_TYPE;
```
Comment (SimenB) [inline example/server.js:96]: should go through `setContentType` I guess?

Comment (voltbit) [reply example/server.js:96]: Thanks, yeah there is a lot of inconsistency in how that flag is set, I'll iron them out.

**Comment by voltbit** [comment]
Hi @zbjornson could you please trigger the tests again? The PR should be ready for review :).

**Comment by zbjornson** [comment]
CI is green! I'll try to review this this weekend and hopefully @SimenB and/or @siimon can also review soon.

**Comment by shyimo** [comment]
thanks @voltbit & @zbjornson! we are excited about this feature!

**Comment by shyimo** [comment]
Hi @zbjornson.
Any news on that ?

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,15 @@ project adheres to [Semantic Versioning](http://semver.org/).
...
 ### Changed
 
+- feat: support for OpenMetrics registries, including exemplars
+
+- feat: out of the box support for exemplars populated with telemetry data on
+  default metrics when OpenTelemetry is loaded in the project
+
+- feat: new option for calling `observe()` and `inc()` methods on the histogram
+  and counter metric types that can be passed an object of format
+  `{labels: a, value: b, exemplarLabels: c}`
```
Comment (SimenB) [inline CHANGELOG.md:21]: should this be a separate PR?

File: README.md [hunk 3]
```diff
@@ -390,6 +390,47 @@ Default labels will be overridden if there is a name conflict.
...
+and Histogram metric types. The default metrics have support for OpenTelemetry,
+they will populate the exemplars with the labels `{traceId, spanId}` and their
+corresponding values.
+
+The format for `inc()` and `observe()` calls are different if exemplars are
+enabled. They get a single object with the format
+`{labels, value, exemplarLabels}`.
+
+When using exemplars, the registry used for metrics should be set to OpenMetrics
+type (including the global or default registry).
```
Comment (SimenB) [inline README.md:405]: can we detect if this isn't done and throw a helpful error?

File: README.md [hunk 4]
```diff
@@ -404,6 +445,9 @@ Registry has a `merge` function that enables you to expose multiple registries
 on the same endpoint. If the same metric name exists in both registries, an
 error will be thrown.
 
+Merging registries of different types is undefined. The user needs to make sure
+all used registries have the same type (Prometheus or OpenMetrics versions).
```
Comment (SimenB) [inline README.md:449]: again, can we throw a usweful error?

File: index.d.ts [hunk 5]
```diff
@@ -66,6 +70,13 @@ export class Registry {
 	 */
 	contentType: string;
 
+	/**
+	 * Set the content type of a registry. Used to change between Prometheus and
+	 * OpenMetrics versions.
+	 * @param contentType The type of the registry
+	 */
+	setContentType(contentType: string): void;
```
Comment (SimenB) [inline index.d.ts:78]: should we type this stricter? I.e. only allow the two known values?

File: index.d.ts [hunk 6]
```diff
@@ -80,10 +91,21 @@ export type Collector = () => void;
...
 
+/**
+ * HTTP Prometheus Content-Type for metrics response headers.
+ */
+export const prometheusContentType: string;
+
+/**
+ * HTTP OpenMetrics Content-Type for metrics response headers.
+ */
+export const openMetricsContentType: string;
```
Comment (SimenB) [inline index.d.ts:107]: same here - instead of `string` the actual string as constant?

File: lib/counter.js [hunk 7]
```diff
@@ -4,20 +4,53 @@
...
+		this.defaultExemplarLabelSet = {};
+		if (config.enableExemplars) {
+			this.enableExemplars = true;
+			this.inc = this.incWithExemplar;
+		} else {
+			this.inc = this.incWithoutExemplar;
+		}
+	}
+
+	enableExemplars(enable = false) {
```
Comment (SimenB) [inline lib/counter.js:33]: are these `enableExemplars` methods used? what's the use case for them?

File: test/metrics/maxFileDescriptorsTest.js [hunk 8]
```diff
@@ -1,29 +1,37 @@
...
+		register.setContentType(regType);
+	});
+
 	afterEach(() => {
 		register.clear();
 	});
 
 	if (process.platform !== 'linux') {
-		it('should not add metric to the registry', async () => {
+		it('should not add metric to the $tag registry', async () => {
```
Comment (SimenB) [inline test/metrics/maxFileDescriptorsTest.js:26]: this is not interpolated, you need to do something like

```suggestion
		it(`should not add metric to the ${tag} registry`, async () => {
```

manually 🙂

**Review by SimenB — COMMENTED** [review]
skimmed through, looks good! Lots of code, but existent tests seems to pass, so safe enough?

**Comment by dnutels** [comment]
Hi guys. Are there, by chance, any updates on how this is progressing and an estimate, hopefully?

**Comment by voltbit** [comment]
> Hi guys. Are there, by chance, any updates on how this is progressing and an estimate, hopefully?

Hi @dnutels  I will start working again on this in a couple of weeks, I will implement the changes requested around mid May, but I cant work on it earlier.

**Comment by voltbit** [comment]
New runs for benchmarks with the latest changes.

<details>
  <summary>Expand for benchmark results</summary>

### Summary:

✗ registry ➭ getMetricsAsJSON#1 with 64 is 11.43% slower.
✗ registry ➭ getMetricsAsJSON#2 with 8 is 155.0% slower.
✓ registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 is 58.48% faster.
✗ registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 is 23.80% slower.
⚠ registry ➭ getMetricsAsJSON#6 with 2 is 6.297% acceptably slower.
⚠ registry ➭ metrics#1 with 64 is 5.984% acceptably slower.
⚠ registry ➭ metrics#2 with 8 is 1.545% acceptably slower.
⚠ registry ➭ metrics#2 with 4 and 2 with 2 is 0.09495% acceptably slower.
✓ registry ➭ metrics#2 with 2 and 2 with 4 is 35.38% faster.
⚠ registry ➭ metrics#6 with 2 is 3.828% acceptably slower.
✓ histogram ➭ observe#1 with 64 is 0.2496% faster.
⚠ histogram ➭ observe#2 with 8 is 0.8472% acceptably slower.
⚠ histogram ➭ observe#2 with 4 and 2 with 2 is 0.7701% acceptably slower.
✓ histogram ➭ observe#2 with 2 and 2 with 4 is 3.590% faster.
⚠ histogram ➭ observe#6 with 2 is 3.784% acceptably slower.
✓ gauge ➭ inc is 2.087% faster.
✓ gauge ➭ inc with labels is 0.2869% faster.
✓ summary ➭ observe#1 with 64 is 3.924% faster.
✓ summary ➭ observe#2 with 8 is 0.2697% faster.
✓ summary ➭ observe#2 with 4 and 2 with 2 is 1.130% faster.
✓ summary ➭ observe#2 with 2 and 2 with 4 is 0.4797% faster.
✓ summary ➭ observe#6 with 2 is 1.389% faster.

### Summary:

✓ registry ➭ getMetricsAsJSON#1 with 64 is 1.990% faster.
✓ registry ➭ getMetricsAsJSON#2 with 8 is 321.0% faster.
⚠ registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 is 8.586% acceptably slower.
✓ registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 is 13.94% faster.
✗ registry ➭ getMetricsAsJSON#6 with 2 is 16.18% slower.
✓ registry ➭ metrics#1 with 64 is 20.45% faster.
⚠ registry ➭ metrics#2 with 8 is 0.8526% acceptably slower.
✓ registry ➭ metrics#2 with 4 and 2 with 2 is 5.139% faster.
⚠ registry ➭ metrics#2 with 2 and 2 with 4 is 6.645% acceptably slower.
✓ registry ➭ metrics#6 with 2 is 0.7693% faster.
⚠ histogram ➭ observe#1 with 64 is 3.102% acceptably slower.
✓ histogram ➭ observe#2 with 8 is 1.538% faster.
✓ histogram ➭ observe#2 with 4 and 2 with 2 is 0.4875% faster.
✓ histogram ➭ observe#2 with 2 and 2 with 4 is 0.7254% faster.
✓ histogram ➭ observe#6 with 2 is 2.165% faster.
✓ gauge ➭ inc is 22.42% faster.
⚠ gauge ➭ inc with labels is 1.116% acceptably slower.
✓ summary ➭ observe#1 with 64 is 2.565% faster.
⚠ summary ➭ observe#2 with 8 is 0.2203% acceptably slower.
⚠ summary ➭ observe#2 with 4 and 2 with 2 is 0.08091% acceptably slower.
✓ summary ➭ observe#2 with 2 and 2 with 4 is 1.113% faster.
✓ summary ➭ observe#6 with 2 is 1.459% faster.

### Summary:

✗ registry ➭ getMetricsAsJSON#1 with 64 is 12.81% slower.
✗ registry ➭ getMetricsAsJSON#2 with 8 is 563.2% slower.
✓ registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 is 7.191% faster.
⚠ registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 is 8.622% acceptably slower.
✗ registry ➭ getMetricsAsJSON#6 with 2 is 13.17% slower.
✗ registry ➭ metrics#1 with 64 is 25.55% slower.
⚠ registry ➭ metrics#2 with 8 is 2.849% acceptably slower.
⚠ registry ➭ metrics#2 with 4 and 2 with 2 is 3.033% acceptably slower.
⚠ registry ➭ metrics#2 with 2 and 2 with 4 is 2.256% acceptably slower.
⚠ registry ➭ metrics#6 with 2 is 4.078% acceptably slower.
⚠ histogram ➭ observe#1 with 64 is 1.686% acceptably slower.
✓ histogram ➭ observe#2 with 8 is 0.3382% faster.
⚠ histogram ➭ observe#2 with 4 and 2 with 2 is 0.2078% acceptably slower.
⚠ histogram ➭ observe#2 with 2 and 2 with 4 is 6.005% acceptably slower.
⚠ histogram ➭ observe#6 with 2 is 2.634% acceptably slower.
✓ gauge ➭ inc is 12.39% faster.
✓ gauge ➭ inc with labels is 2.485% faster.
✓ summary ➭ observe#1 with 64 is 9374% faster.
✓ summary ➭ observe#2 with 8 is 2.167% faster.
✗ summary ➭ observe#2 with 4 and 2 with 2 is 19.97% slower.
⚠ summary ➭ observe#2 with 2 and 2 with 4 is 1.552% acceptably slower.
⚠ summary ➭ observe#6 with 2 is 0.3165% acceptably slower.
</details>

File: lib/counter.js (hunk 7 above)
Comment (voltbit) [reply lib/counter.js]: I removed them, they were not used and implementing a toggle for exemplars on existing metrics is problematic.

File: CHANGELOG.md (hunk 2 above)
Comment (voltbit) [reply CHANGELOG.md:21]: Right now the implementation is quite tightly coupled to the feature for exemplars. I though about making this more generic from the beginning but I was worried about the performance impact. I tried to provide 100% backwards compatibility and minimal performance impact with the change.

If I am to make a separate PR for it I would make this feature afailable for all metrics (not just counters and histograms) for the sake of consistency and I would 

Right now the call with `inc({value, labels, exemplarLabels})` can only be done if the metric was created with exemplars enabled.

If I make this call more generic - allow inc() and observe() to be called with an object - I would need to differentiate between the `inc({value, labels, exemplarLabels})` and the call that already exists `inc({labelNames})`. This might not be that bad, but it implies a check on the keys of the object every time inc/obs are used.

I am not sure which is the better aproach, let me know if you have feedback about this.

So the alternative to what we have now in the PR is: implement an object type check on all metric types and allow them to receive a single object containing all the relevant data for ins/obs operations. This might be useful on the long run, but it may as well be a feature that slows down the library for years until we need more fields like we do with exemplars (which are kind of a fringe feature anyway).

File: README.md (hunk 4 above)
Comment (voltbit) [reply README.md:449]: https://github.com/siimon/prom-client/pull/482/files#diff-c76cf141522a047fee59ae738cbc8a0dc2c48ed15666d958d407d867c3a4b69fR237

File: README.md (hunk 3 above)
Comment (voltbit) [reply README.md]: https://github.com/siimon/prom-client/pull/482/files#diff-50029c8b3fcc5c34781f42a629d19b95b603a0ab9306ac642e3355e297d0233fR54

File: index.d.ts (hunk 5 above)
Comment (voltbit) [reply index.d.ts]: I have added a stricter typing to the, let me know what you think. I used string template types from typescript in case the content type will have new versions/encodings in the future.

**Comment by shyimo** [comment]
Hi @zbjornson. any news regarding the PR ?

**Comment by hitwill** [comment]
Bumping for interest

**Comment by vjsamuel** [comment]
Bumping for interest again :)

**Comment by isaac-elvt** [comment]
Bumping for interest as well, this would save my organization so much pain!

**Comment by xal3xhx** [comment]
bump for intrest. i really need this

**Comment by shyimo** [comment]
bump for interest. really important task

**Comment by ejba** [comment]
bumping for interest!

File: README.md (hunk 4 above)
Comment (SimenB) [reply README.md:449]: Docs should be updated to say it throws, no? Not undefined behavior

File: example/exemplars.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
+import * as prom from '../index';
```
Comment (SimenB) [inline example/exemplars.ts:1]: I don't think we need both a JS and a TS example - could you remove this one?

File: index.d.ts [hunk 10]
```diff
@@ -150,16 +185,29 @@ interface MetricConfiguration<T extends string> {
 	name: string;
 	help: string;
 	labelNames?: T[] | readonly T[];
-	registers?: Registry[];
+	registers?: Registry<PrometheusContentType | OpenMetricsContentType>[];
```
Comment (SimenB) [inline index.d.ts:188]:
```suggestion
	registers?: (Registry<PrometheusContentType> | Registry<OpenMetricsContentType>)[];
```

right? A registry itself cannot contain either

File: lib/cluster.js [hunk 11]
```diff
@@ -84,19 +84,30 @@ class AggregatorRegistry extends Registry {
 		});
 	}
 
+	get contentType() {
```
Comment (SimenB) [inline lib/cluster.js:87]: is this needed? won't the class just inherit from `super`?

File: lib/histogram.js [hunk 12]
```diff
@@ -27,14 +43,21 @@ class Histogram extends Metric {
 			return acc;
 		}, {});
 
+		this.bucketExemplars = this.upperBounds.reduce((acc, upperBound) => {
```
Comment (SimenB) [inline lib/histogram.js:46]: should this be created even if `config.enableExemplars` isn't `true`?

File: package.json [hunk 13]
```diff
@@ -41,10 +41,11 @@
 		"jest": "^26.0.1",
 		"lint-staged": "^10.0.4",
 		"nock": "^13.0.5",
-		"prettier": "2.0.5",
+		"prettier": "2.6.2",
```
Comment (SimenB) [inline package.json:44]: this should be done on master, not this PR

**Review by SimenB — APPROVED** [review]
this generally LGTM. Would love another set of eyes, tho.

Thanks for a fantastic contribution, @voltbit!

File: package.json (hunk 13 above)
Comment (SimenB) [reply package.json:44]: hah, I actually updated some time ago: https://github.com/siimon/prom-client/pull/510

so this just needs a rebase/merge of `master`

**Comment by SimenB** [comment]
@zbjornson @siimon PTAL 🙂

**Comment by voltbit** [comment]
Thanks for all the feedback and all the interest shown! I'll rebase and check the comments as soon as possible (this weekend most likely).

**Comment by ejba** [comment]
@voltbit is there anything we can do to help you?

File: CHANGELOG.md (hunk 2 above)
Comment (SimenB) [reply CHANGELOG.md]: >So the alternative to what we have now in the PR is: implement an object type check on all metric types and allow them to receive a single object containing all the relevant data for ins/obs operations. This might be useful on the long run, but it may as well be a feature that slows down the library for years until we need more fields like we do with exemplars (which are kind of a fringe feature anyway).

One thing we could do is drop support for passing `labels` (`inc({labelNames})` like you mention). That way you either pass a single value or an object. I don't `typeof arg === 'number'` adds much overhead, and then the object shape would be what we want.

Would require a major version of course, but that's not an issue. We wanna drop old nodes anyways

File: lib/metrics/processCpuTotal.js [hunk 14]
```diff
@@ -1,6 +1,8 @@
 'use strict';
 
+const OtelApi = require('@opentelemetry/api');
```
Comment (SimenB) [inline lib/metrics/processCpuTotal.js:3]: this should be added to `package.json`

File: lib/registry.js [hunk 15]
```diff
@@ -10,27 +11,44 @@ function escapeLabelValue(str) {
...
+
+	constructor(regContentType = Registry.PROMETHEUS_CONTENT_TYPE) {
 		this._metrics = {};
 		this._collectors = [];
 		this._defaultLabels = {};
+		if (
+			regContentType !== Registry.PROMETHEUS_CONTENT_TYPE &&
+			regContentType !== Registry.OPENMETRICS_CONTENT_TYPE
+		) {
+			throw new TypeError('Content type unsupported');
```
Comment (SimenB) [inline lib/registry.js:38]:
```suggestion
			throw new TypeError(`Content type ${regContentType} is unsupported`);
```

**Comment by voltbit** [comment]
hi @shyimo & @ejba & other contributors
I am not sure I will get the time to work on this again before the holiday season. If this work is urgent for you and you want to pick up the task feel free to do so.

**Comment by vothanhbinhlt** [comment]
Hello team, any update? I am waiting this feature

**Comment by SimenB** [comment]
There was literally an update in the post before yours

**Comment by vothanhbinhlt** [comment]
> There was literally an update in the post before yours

I see the PR is draft status. Is it merged to the main branch and publish to a new version?

**Comment by SimenB** [comment]
Superseded by #544
